### ISSUE TITLE
Nimbus set-up for Onboarding experiment

### ISSUE THREAD

**Issue by amedyne** [mozilla-mobile/focus-android#6387]
Feature flag: #6297 

Feature should be controlled by a Nimbus experiment in preparation for the planned onboarding experiment.
Experiment brief [here](https://docs.google.com/document/d/1XzTeAqOYR2McZrtaqqbf1ZCMZKTB5aFKOeb42kPNZTQ/edit).

**Comment by amedyne** [comment]
Based on the discussion yesterday, the experiment branches/scenarios will now be:
1.) current onboarding experience, 
2.) New onboarding experience: new onboarding card + all CFRs
3.) Onboarding card only (without the CFRs)
Because of this we need to separate Nimbus Feature for the onboarding card and the CFRs.

**Comment by ionutbedregeanu** [comment]
To start working on this ticket we need to have all new features merged, so I will move it back to Backlog until then

**Comment by ionutbedregeanu** [comment]
@amedyne I started to work on this but from what I see we don't have the experiments.
I will make a draft which will be finished when the experiments will be added

**Comment by jhugman** [comment]
Nimbus requires a network fetch before it can enroll in any experiments (it needs to download the experiment recipes).

This means running experiments in early startup needs to handle startup slightly differently during the app's first start-up.

You should find a place in your app in your first launch startup where:
 * time has elapsed enough for a network request
 * it is safe to call `applyPendingExperiments()`.

Before this point, you won't be able to do any experiments— because there will never be any experiments downloaded.

An ideal point might be as the user moves from the first screen of the onboarding flow.

Calling `getVariables()` (or `FocusNimbus.features.onboarding.value()`) either side of the `applyPendingExperiments()` is not guaranteed to be the same value.

```kotlin
// At first startup, the experiment recipes haven't downloaded,
// so this is always the defaults shipped with the app.
val config1 = FocusNimbus.features.onboarding.value()
val config2 = FocusNimbus.features.onboarding.value()
assert(config1 == config2)

// At the checkpoint: we've probably had enough time to download
// the experiment recipes.
nimbus.applyPendingExperiments()

val config3 = FocusNimbus.features.onboarding.value()
val config4 = FocusNimbus.features.onboarding.value()
assert(config3 == config4)

// The user has seed the onboarding flow.
// Nimbus will check if we're in an experiment and, if so, send an exposure event.
FocusNimbus.features.onboarding.recordExposure()
```

What we cannot guarantee is that `config3 == config1`. Between calls to `applyPendingExperiments()` we can make some guarantees around idempotency.

I see you've put this `applyPendingExperiments()` [call to immediately follow the `fetch`][1]. This is problematic for two reasons: 

[1]: https://github.com/mozilla-mobile/focus-android/blob/c3ea111f6f2a95f6415b62fd1b058348b1315f8c/app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt#L87-L97

1. the apply-pending-experiments call happens at a non-deterministic place some place in the startup sequence, so if you're calling into nimbus for the same feature at multiple points (not unlikely, especially for a multi-page onboarding feature), 
  - your onboarding will likely be misconfigured, and your experimental configuration will not reflect what the onboarding flow that the users will see.
  - worse, you will not detect this breakage unless you're doing testing the onboarding flow within an experiment *AND* on a slow/high-latency network connection.
2. when the app records exposure events for `onboarding`, Nimbus won't know if the user has seen some, all or none of the experimental treatment.
  - from a data science point of view, this will potentially render your experimentation on the feature meaningless :(
  - worse! since this is being done on _every_ startup, this is going to affect every feature that is used close to startup.

I'd be super-happy to help talk this through or pair with you if you need some help or have further questions.

/cc @ionutbedregeanu

**Comment by ionutbedregeanu** [comment]
@Amejia481 @amedyne @jhugman The onboarding screen is the first screen which should be shown when the app is opened for the first time and we don't have enough time to download the experiments and decide which onboarding experiment should we apply.
I asked @Mugurell and @mcarare for suggestions and they give me two possible approaches:
- fetch the experiments in the background in order to already have it when we are opening the app
- decide locally(without nimbus experiments) which onboarding branch should we follow and based on the telemetry sent after that the DS can see how many users are part of a specific experiment

**Comment by amedyne** [comment]
@ionutbedregeanu What is the impact to the app start-up when we fetch the experiments in the background after install?

**Comment by ionutbedregeanu** [comment]
If we try to fetch the experiments in a background service and for some reasons(e.g poor internet conenction or is taking too much) the experiments aren't downloaded we are constrained to use the default values, which may be different of those from nimbus api.
As Arturo said, if we want to decide which onboarding branch we need to follow based on nimbus api results we need to add some UI to indicate that we are loading, which in my opinion it will impact also the user experience
Also this approach is more complex than the second one

**Comment by ionutbedregeanu** [comment]
Sounds good to me. I will put ticket on waiting until UX also approves this approach and help us with some design @jeffreygee ^^

**Comment by Amejia481** [comment]
To find a good time limit for the splash screen, we did some testing to see how long it takes in average to download the experiments in Firefox release:

**Good network**
0.097 seconds

**Bad network**
0.744 seconds

We believe that a good time limit to start of could be max/min **2 seconds**, and we think **we should show the screen also for at least for 2 seconds to avoid abrupt flashes**, in other words, the splash screen should always be shown for 2 second, if we were not able to download the experiments we will show the default experience.

**Comment by mcarare** [comment]
IMO we should consider adding a probe to measure how many times the "timeout" threshold has been reached. It could serve as a benchmark for other future similar experiments (maybe even Fenix onboarding).

**Comment by Amejia481** [comment]
Yeah, ideally we would like to do something pretty simple, a screen with just the logo and saying we are initializing Focus it could be enough.

**Comment by Amejia481** [comment]
On Android 12, the OS shows a default splash screen, that could give us a hint. It's only a background with the logo in the middle.

**Comment by ionutbedregeanu** [comment]
@heres-maria ^^ This is ready to be tested.
Additional tickets started from this ticket, which are in progress or not implemented yet: https://github.com/mozilla-mobile/focus-android/issues/6847 https://github.com/mozilla-mobile/focus-android/issues/6939

**Comment by heres-maria** [comment]
The following issues were filed so far during bugfix verification: #7069, #7070, #7082, #7083, #7085, #7087.

**Comment by amedyne** [comment]
Notes from discussion today: 
Is there a way to not enroll users that have "problems" or delay with downloading the experiment recipe?
